File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.*;

class Solution {
    public int solution(String s) {
        int answer = 0;
        
        for(int i = 0; i < s.length(); i++){        
            if(isValidString(s)) answer++;
            s = rotateString(s);
        }
        
        return answer;
    }
    
    public boolean isValidString(String str){
        Stack<Character> stack = new Stack<>();
        for(char ch: str.toCharArray()){
            if(ch == '(' || ch == '{' || ch == '['){
                stack.add(ch);
            } else{
                if(stack.isEmpty()) return false;
                
                if(ch == ')' && stack.peek() == '(') stack.pop();
                else if(ch == '}' && stack.peek() == '{') stack.pop();
                else if(ch == ']' && stack.peek() == '[') stack.pop();
                else break;
            }
        }
        return stack.isEmpty();
    }
    
    public String rotateString(String str){
        return str.substring(1, str.length()) + str.charAt(0);
    }
}

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
def solution(s):
    def check_valid(tgt_str):
        stack = []
        for ch in tgt_str:
            if ch == '(' or ch == '{' or ch == '[':
                stack.append(ch)
            elif ch == ')':
                if not stack:
                    return False
                if stack and stack[-1] == '(':
                    stack.pop()
            elif ch == '}':
                if not stack:
                    return False
                if stack and stack[-1] == '{':
                    stack.pop()
            elif ch == ']':
                if not stack:
                    return False
                if stack and stack[-1] == '[':
                    stack.pop()
        return not stack
    
    str_len = len(s)
    str_list = list(s)
    answer = 0
    
    for i in range(str_len):
        if check_valid(str_list):
            answer += 1
        str_list.append(str_list.pop(0))
    
    return answer